File: examples/arty_managed/oled_demo/oled_demo.py
# -*- coding: utf-8 -*-

# Copyright 2023 The Aerospace Corporation
#
# This file is part of SatCat5.
#
# SatCat5 is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the
# Free Software Foundation, either version 3 of the License, or (at your
# option) any later version.
#
# SatCat5 is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
# License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with SatCat5.  If not, see <https://www.gnu.org/licenses/>.

"""
Console demo for remote control of the "arty_managed" OLED

The application opens the designated UART interface, then begins sending
a repeating sequence of updates to the OLED display.
"""

# Python library imports
import logging, os, sys, time
from datetime import datetime
from traceback import format_exception

# Additional imports from SatCat5 core.
sys.path.append(os.path.join(
    os.path.dirname(__file__), '..', '..', '..', 'src', 'python'))
import satcat5_cfgbus
import satcat5_i2c
import satcat5_oled
import satcat5_uart

# Start logging system.
logger = logging.getLogger(__name__)
logger.info('Starting logger...')
logger.setLevel('INFO')

# MAC addresses for the Ethernet-over-UART interface.
REMOTE_MAC = b'\xDE\xAD\xBE\xEF\xCA\xFE'

# ConfigBus address for the I2C controller:
DEVADDR_I2C = 10

class ArtyDemo:
    def __init__(self, if_obj):
        """Create a demo controller linked to the designated interface."""
        # Create each of the driver objects:
        self.cfg = satcat5_cfgbus.ConfigBus(if_obj, REMOTE_MAC)
        self.i2c = satcat5_i2c.I2cController(self.cfg, DEVADDR_I2C)
        self.oled = satcat5_oled.Ssd1306(self.i2c)
        # Counter cycles between display modes.
        self.cycle = 0

    def ok(self):
        """Connectivity test."""
        regval = self.cfg.read_reg(DEVADDR_I2C, 0)
        return regval is not None

    def display_next(self):
        """Write the next message to the OLED display."""
        # Format the current time.
        day = datetime.now().strftime('%Y-%m-%d')
        now = datetime.now().strftime('%H:%M:%S')
        # Generate the complete message...
        if self.cycle == 0:
            msg = f'Time: {now:9s} Date: {day}'
        elif self.cycle == 1:
            msg = f'Time: {now:9s} SatCat5 demo!'
        else:
            msg = f'Time: {now:9s} Meow meow meow.'
        # Update message index and write to screen.
        ok = self.oled.display(msg)
        if ok:
            self.cycle = (self.cycle + 1) % 3
            logger.info(msg)
        else:
            logger.error('Error updating OLED.')
        return ok

    def run_forever(self):
        """Write a new message every second until user hits Ctrl+C."""
        while self.display_next():
            time.sleep(1.0)

# Main function: Initialize and run the demo.
if __name__ == '__main__':
    # Configure logging to print to console.
    logger.addHandler(logging.StreamHandler(sys.stdout))
    # User must specify the UART port to select.
    if len(sys.argv) > 1:
        uart_name = sys.argv[1]
    else:
        print('Usage: python3 oled_demo.py [uart_name]')
        print('  Where "ifname" is the USB-UART attached to the arty_managed FPGA.')
        sys.exit(1)
    # Start application and catch any errors.
    try:
        port = satcat5_uart.AsyncSLIPPort(uart_name, logger)
        demo = ArtyDemo(port)
        if demo.ok():
            demo.run_forever()
        else:
            logger.error('Could not connect to I2C interface.')
            sys.exit(1)
    except KeyboardInterrupt:
        sys.exit(0)
    except Exception as e:
        logger.error("Error: %s" % format_exception(e))
        sys.exit(1)

File: examples/arty_managed/oled_demo/oled_demo.cc
//////////////////////////////////////////////////////////////////////////
// Copyright 2023 The Aerospace Corporation.
// This file is a part of SatCat5, licensed under CERN-OHL-W v2 or later.
//////////////////////////////////////////////////////////////////////////
// Console demo for remote control of the "arty_managed" OLED
//
// The application opens the designated UART interface, then begins sending
// a repeating sequence of updates to the OLED display.
//

#include <clocale>
#include <cstdio>
#include <ctime>
#include <iostream>
#include <string>
#include <hal_devices/i2c_ssd1306.h>
#include <hal_posix/posix_uart.h>
#include <hal_posix/posix_utils.h>
#include <satcat5/cfgbus_i2c.h>
#include <satcat5/cfgbus_remote.h>

using namespace satcat5;

// Global background services.
log::ToConsole logger;          // Print Log messages to console
util::PosixTimekeeper timer;    // Link system time to internal timers

// MAC addresses for the Ethernet-over-UART interface.
eth::MacAddr LOCAL_MAC  = {0x12, 0x34, 0x56, 0x78, 0x9A, 0xBC};
eth::MacAddr REMOTE_MAC = {0xDE, 0xAD, 0xBE, 0xEF, 0xCA, 0xFE};

// ConfigBus address for the I2C controller:
const unsigned DEVADDR_I2C = 10;

// Configuration of ZCU208 design.
class ArtyDemo final : public poll::Timer
{
public:
    explicit ArtyDemo(cfg::ConfigBus* cfg)
        : m_cycle(0)
        , m_cfg(cfg)
        , m_i2c(cfg, DEVADDR_I2C)
        , m_oled(&m_i2c)
    {
        // Disable interrupts to avoid disrupting the Microblaze CPU.
        // (Since it doesn't know we're remote-controlling this interface.)
        m_i2c.irq_disable();

        // Update screen once per second.
        timer_every(1000);
    }

    // Connectivity test
    bool ok()
    {
        u32 rdval;
        unsigned regaddr = m_cfg->get_regaddr(DEVADDR_I2C, 0);
        return m_cfg->read(regaddr, rdval) == cfg::IOSTATUS_OK;
    }

    // Service loop including simulated interrupt.
    void service()
    {
        m_i2c.request_poll();
        poll::service_all();
    }

private:
    // Timer event handler.
    void timer_event() override {
        // Buffer = 16x2 characters + NULL termination
        char day[11], now[9], msg[33];

        // Format the current time.
        time_t rawtime;
        struct tm* timeinfo;
        time(&rawtime);
        timeinfo = localtime(&rawtime);
        strftime(day, sizeof(day), "%Y-%m-%d", timeinfo);
        strftime(now, sizeof(now), "%H:%M:%S", timeinfo);

        // Generate the complete message...
        switch (m_cycle) {
            case 0:
                snprintf(msg, sizeof(msg), "Time: %9s Date: %s", now, day);
                break;
            case 1:
                snprintf(msg, sizeof(msg), "Time: %9s SatCat5 demo!", now);
                break;
            case 2:
                snprintf(msg, sizeof(msg), "Time: %9s Meow meow meow.", now);
                break;
        }

        // Update message index and write to screen.
        if (m_oled.display(msg)) {
            m_cycle = (m_cycle + 1) % 3;
            std::cout << msg << std::endl;
        }
    }

    // Counter cycles between various messages.
    unsigned m_cycle;

    // Remotely-controlled ConfigBus peripherals.
    cfg::ConfigBus* const m_cfg;
    cfg::I2c m_i2c;
    device::i2c::Ssd1306 m_oled;
};

void oled_demo(io::SlipUart* uart)
{
    // Open remote-control interface.
    auto dispatch = new eth::Dispatch(LOCAL_MAC, uart, uart);
    auto cfgbus = new eth::ConfigBus(dispatch, timer.timer());
    cfgbus->connect(REMOTE_MAC);
    cfgbus->set_timeout_rd(200000);

    // Attach the OLED driver.
    auto oled = new ArtyDemo(cfgbus);

    // Poll until communications fail or user hits Ctrl+C.
    poll::service_all();
    while (oled->ok()) {
        util::sleep_msec(10);
        oled->service();
    }

    delete oled;
    delete cfgbus;
    delete dispatch;
}

int main(int argc, const char* argv[])
{
    // Set console mode for UTF-8 support.
    setlocale(LC_ALL, SATCAT5_WIN32 ? ".UTF8" : "");

    // Parse command-line arguments.
    std::string ifname;
    unsigned baud = 921600;
    if (argc == 3) {
        ifname = argv[1];       // UART device
        baud = atoi(argv[2]);   // Baud rate
    } else if (argc == 2) {
        ifname = argv[1];       // UART device
    }

    // Print the usage prompt?
    if (argc > 3 || ifname == "" || ifname == "help" || ifname == "--help") {
        std::cout << "oled_demo uses arty_managed to control an OLED screen." << std::endl
            << "Usage: oled_demo.bin <ifname>" << std::endl
            << "       oled_demo.bin <ifname> <baud>" << std::endl
            << "Where 'ifname' is the USB-UART attached to the arty_managed FPGA." << std::endl;
        return 0;
    }

    // Open the specified UART interface.
    io::SlipUart* uart = new io::SlipUart(ifname.c_str(), baud);

    // Interface ready?
    if (uart && uart->ok()) {
        std::cout << "Starting oled_demo on " << ifname << std::endl;
        oled_demo(uart);
        return 0;
    } else {
        std::cerr << "Couldn't open UART interface: " << ifname << std::endl;
        return 1;
    }
}
